pkg/controller/operator/master/resources/kubermatic/common.go

```diff
@@ -112,22 +112,41 @@ func IngressCreator(cfg *operatorv1alpha1.KubermaticConfiguration) reconciling.N
 				ServicePort: intstr.FromInt(80),
 			}
 
+			// PathType has been added in Kubernetes 1.18 and defaults to
+			// "ImplementationSpecific". To prevent reconcile loops in previous
+			// Kubernetes versions, this code is carefully written to not
+			// overwrite a PathType that has been defaulted by the apiserver.
+			var pathType *extensionsv1beta1.PathType
+
+			// As we control the entire rule set anyway, it's enough to find
+			// the first pathType -- they will all be identical eventually.
+			if rules := i.Spec.Rules; len(rules) > 0 {
```

Comment:
This should make it, but I would suggest assigning the current rules to a new variable and getting the relevant pathType with something like:

```
func getPathType(host string, path string, rules []extensionsv1beta1.IngressRule) *extensionsv1beta1.PathType {
	for _, rule := range rules {
		if http := rule.IngressRuleValue.HTTP; rule.Host == host && http != nil {
			for _, p := range http.Paths {
				if p.Path == path {
					return p.PathType
				}
			}
		}
	}
	return nil
}
```

I find it clearer, a bit more robust and future-proof.

Reply:
I don't really understand this suggestion. What about your code is clearer than mine? Because it's in a dedicated function? Because you (IMHO needlessly) compare the host?

Reply:
Because it tries to match the `PathType` associated to the same `Rule`, `Path` and not just taking the first `Path` of the first `Rule`. For instance if a rule is added on top of the manifest, it will be removed by the reconcile logic, but its PathType will be preserved and kept for all paths. This scenario is quite remote. Second scenario, is if in future we want to have different TypePath values we can keep this logic, as it is slightly more generic while your is tailored to this specific use case.

What I had in mind is something like:
```
			currRules := i.Spec.Rules
			i.Spec.Rules = []extensionsv1beta1.IngressRule{
				{
					Host: cfg.Spec.Ingress.Domain,
					IngressRuleValue: extensionsv1beta1.IngressRuleValue{
						HTTP: &extensionsv1beta1.HTTPIngressRuleValue{
							Paths: []extensionsv1beta1.HTTPIngressPath{
								{
									Path:     "/api",
									PathType: getPathType("", "/api", currRules),
									Backend: extensionsv1beta1.IngressBackend{
										ServiceName: apiServiceName,
										ServicePort: intstr.FromInt(80),
									},
								},
								{
									Path:     "/",
									PathType: getPathType("", "/api", currRules),
									Backend:  *i.Spec.Backend,
								},
							},
						},
					},
				},

```

I agree that the gain is small, but the added complexity is small too. I'm not going to be pushy on that, take it as a suggestion.

Reply:
Let's burn that bridge when we come to it. I can see the point of keeping the "wrong" PathType in some exotic situations, but this might also be seen as a feature :-D

Reply:
Well, having two paths with different types in the same ingress does not seem to be that "exotic", anyway it's your call.

Reply:
Having two path types in *this particular* Ingress that is managed by the operator and therefore not the place where anyone can feasably add their own paths or path types, is indeed something I would call exotic ;-)

I had long discussions with previous project members about how much we want to "allow" others to modify the resources we reconcile (e.g. should the operator remove unexpected volume mounts on a deployment?) and the consensus was that for "our" resources, we do not want interference and will not "be kind" to other modifications. This means that the KKP operator does not make any statement over what modifications will survive and users should generally never touch "our" resources. This is the assumption under which all reconcile functions are written in KKP. Just for a little background. :)